File: hw01/redwards-hw01/rabbits-and-coyotes.R
#
# Reginald Edwards
# 9 April 2012
# reginald.edwards@gmail.com
#
# This script produces a plot of the population dynamics of the rabbit-coyote
# system. It relies on input data from "rabbits-and-coyotes.cpp"
#

population <- read.table("population.txt", header=FALSE, sep=" ")

names(population) <- c("t", "rabbits", "coyotes")

y.range <- range(population$rabbits, population$coyotes)

jpeg("Rabbits and Coyotes.jpg")
  plot(population$t, population$rabbits, type="l", col="red", lwd=4, ylim=y.range, xlab="t", ylab="Population", main="Rabbit and coyote populations over time")
  points(population$t, population$coyotes, type="l", col="blue", lwd=4)
  legend(5000,3, c("rabbits", "coyotes"), col=c("red","blue"), lty=1, lwd=4)
dev.off()

File: hw01/redwards-hw01/rabbits-and-coyotes.cpp
/* 
 Reginald Edwards
 9 April 2012
 reginald.edwards@gmail.com
 
 This program takes as input values constants to a differential equation
 governing the population dynamics of a closed system.
 
 This program produces a space-separated text file of rabbit and coyote 
 populations over time.
*/

#include <iostream>
#include <cstdlib>
using std::cout;
using std::cin;
using std::endl;

// function main begins program execution
int main(int argc, const char* argv[]){

  float alpha, beta, gamma, pi, r, c, N, h;
  
  alpha = atof( argv[1] );
  beta = atof( argv[2] );
  gamma = atof( argv[3] );
  pi = atof( argv[4] );
  r = atof( argv[5] );
  c = atof( argv[6] );
  N = atof( argv[7] ); // number of time periods
  h = atof( argv[8] ); // step size for Euler method
  
  // Use first order Euler approximation
  // see http://en.wikipedia.org/wiki/Euler_method
  float r1, c1;
  int tx=0;
  while(tx<N){
    r1 = r + h*(alpha*r - beta*r*c);
    c1 = c + h*(-1*gamma*c + pi*r*c);
    r = r1;
    c = c1;
    cout << tx << " " << r << " "<< c << endl;
    tx++;
  }
    
}
